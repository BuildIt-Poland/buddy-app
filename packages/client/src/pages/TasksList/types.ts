import { RouteComponentProps } from 'react-router-dom';

export type TasksListProps = Pick<RouteComponentProps, 'history'>;
