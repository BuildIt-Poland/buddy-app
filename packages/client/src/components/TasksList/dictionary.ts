interface TaskListDictionary {
  BUDDY_TAB_TITLE: string;
  NEWBIE_TAB_TITLE: string;
  SUCCESS_MESSAGE: string;
  ERROR_MESSAGE: string;
  PLUS_BUTTON_TITLE: string;
}

const TASK_LIST_DICTIONARY: TaskListDictionary = {
  BUDDY_TAB_TITLE: 'My tasks',
  NEWBIE_TAB_TITLE: 'Newbie Tasks',
  SUCCESS_MESSAGE: 'Task status updated',
  ERROR_MESSAGE: 'An error ocurred updating task',
  PLUS_BUTTON_TITLE: 'Add new task',
};

export default TASK_LIST_DICTIONARY;
