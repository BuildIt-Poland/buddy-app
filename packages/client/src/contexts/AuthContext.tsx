import React, { useReducer, useEffect, useContext } from 'react';
import { Mutation, AuthPayload } from '@buddy-app/schema';
import { useQuery } from '@apollo/react-hooks';
import { ExecutionResult } from '@apollo/react-common';
import { auth } from 'utils';
import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { GET_LOCAL_AUTH } from 'graphql/get-auth.graphql';
import { AuthCache, setCacheToken } from 'utils/apollo-client/cache';
import { apolloClient } from 'utils';

enum ActionTypes {
  AUTH_INIT = 'auth/init',
  AUTH_ERROR = 'auth/error',
  AUTH_SUCCESS = 'auth/success',
  AUTH_LOGOUT = 'auth/logout',
}

interface Action {
  type: ActionTypes;
  payload?: AuthPayload;
}

export interface AuthState {
  loading: boolean;
  error?: any;
  data: AuthPayload;
  isAuthenticated: boolean;
}

const defaultState: AuthState = {
  loading: false,
  error: null,
  data: {} as AuthPayload,
  isAuthenticated: false,
};

interface AuthContext extends AuthState {
  login: (email: string, password: string) => Promise<ExecutionResult<Mutation>>;
  logout: () => void;
}

interface AuthProviderProps {
  children: React.ReactNode;
  value?: AuthContext;
}

const defaultContext: AuthContext = {
  ...defaultState,
  login: () => Promise.reject(),
  logout: () => null,
};

const AuthContext = React.createContext(defaultContext);

const authReducer = (state: AuthState, action: Action): AuthState => {
  switch (action.type) {
    case ActionTypes.AUTH_INIT:
      return {
        ...state,
        loading: true,
        error: null,
      };
    case ActionTypes.AUTH_SUCCESS:
      return {
        ...state,
        loading: false,
        data: { ...action.payload } as AuthPayload,
        isAuthenticated: true,
      };
    case ActionTypes.AUTH_ERROR:
      return {
        ...state,
        loading: false,
        error: { ...action.payload },
      };
    case ActionTypes.AUTH_LOGOUT:
      return {
        ...state,
        data: {} as AuthPayload,
        isAuthenticated: false,
      };
    default:
      throw new Error('Not valid action');
  }
};

const AuthProvider = ({ children, ...props }: AuthProviderProps) => {
  const [state, dispatch] = useReducer(authReducer, defaultState);
  const { data } = useQuery<AuthCache>(GET_LOCAL_AUTH);

  useEffect(() => {
    const user = auth.getUser();
    if (user) {
      dispatch({
        type: ActionTypes.AUTH_SUCCESS,
        payload: user,
      });
    }
  }, []);

  useEffect(() => {
    if (data && data.auth && data.auth.tokenHasExpired) {
      logout();
    }
  }, [data]);

  const login = async (email: string, password: string) => {
    dispatch({ type: ActionTypes.AUTH_INIT });

    try {
      const { data } = await apolloClient.mutate<Mutation>({
        mutation: LOGIN_MUTATION,
        variables: {
          email,
          password,
        },
      });

      if (data) {
        auth.setUser(data.login);
        dispatch({
          type: ActionTypes.AUTH_SUCCESS,
          payload: data.login,
        });
      }

      return Promise.resolve({ data });
    } catch (error) {
      dispatch({ type: ActionTypes.AUTH_ERROR, payload: error });
      return Promise.reject(error);
    }
  };

  const logout = () => {
    auth.removeUser();
    apolloClient.writeData({
      data: {
        ...setCacheToken(false),
      },
    });
    dispatch({ type: ActionTypes.AUTH_LOGOUT });
  };

  return (
    <AuthContext.Provider value={{ ...state, login, logout }} {...props}>
      {children}
    </AuthContext.Provider>
  );
};

const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error(`useAuth must be used within a AuthProvider`);
  }
  return context;
};

export { AuthProvider, useAuth };
