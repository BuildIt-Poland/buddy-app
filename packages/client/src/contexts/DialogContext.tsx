import React, { useState, useContext } from 'react';
import AlertDialog from 'atoms/AlertDialog';

interface State {
  isOpen: boolean;
  message: string;
  title: string | undefined;
  onConfirm?: () => void;
}

interface DialogContextData {
  showDialog: (message: string, title?: string, onConfirm?: () => void) => void;
  hideDialog: () => void;
}

const defaultState: State = {
  isOpen: false,
  title: '',
  message: '',
};

const DialogContext = React.createContext<DialogContextData | undefined>(undefined);

interface DialogProviderProps {
  children: React.ReactNode;
  value?: DialogContextData;
}

const DialogProvider = (props: DialogProviderProps): JSX.Element => {
  const [state, setDialog] = useState<State>(defaultState);

  const showDialog = (message: string, title?: string, onConfirm?: () => void) =>
    setDialog({ isOpen: true, message, title, onConfirm });

  const hideDialog = () => setDialog(defaultState);

  const onConfirmHandler = () => {
    state.onConfirm && state.onConfirm();
    hideDialog();
  };

  return (
    <DialogContext.Provider
      value={{
        showDialog,
        hideDialog,
      }}>
      <AlertDialog
        isOpen={state.isOpen}
        message={state.message}
        onConfirm={state.onConfirm ? onConfirmHandler : undefined}
        title={state.title}
        onClose={hideDialog}
      />
      {props.children}
    </DialogContext.Provider>
  );
};

const useDialog = () => {
  const context = useContext(DialogContext);
  if (context === undefined) {
    throw new Error(`useDialog must be used within a DialogProvider`);
  }
  return context;
};

export { DialogProvider, useDialog };
