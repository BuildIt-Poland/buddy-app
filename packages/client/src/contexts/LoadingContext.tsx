import React, { useState, useContext } from 'react';

interface State {
  loading: boolean;
}

interface LoadingProviderProps {
  children: React.ReactNode;
  value?: State;
}

interface LoadingContextData extends State {
  showLoading: () => void;
  hideLoading: () => void;
}

const defaultState: State = {
  loading: false,
};

const LoadingContext = React.createContext<LoadingContextData | undefined>(
  undefined
);

const LoadingProvider = ({ children, value }: LoadingProviderProps): JSX.Element => {
  const [{ loading }, setLoadingState] = useState<State>(value || defaultState);

  const showLoading = () => setLoadingState(state => ({ ...state, loading: true }));

  const hideLoading = () => setLoadingState(state => ({ ...state, loading: false }));

  return (
    <LoadingContext.Provider
      value={{
        loading,
        showLoading,
        hideLoading,
      }}>
      {children}
    </LoadingContext.Provider>
  );
};

const useLoading = () => {
  const context = useContext(LoadingContext);

  if (context === undefined) {
    throw new Error(`useLoading must be used within a LoadingContext`);
  }
  return context;
};

export { LoadingProvider, useLoading };
