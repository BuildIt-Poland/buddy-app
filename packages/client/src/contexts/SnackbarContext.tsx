import React, { useState, useContext } from 'react';
import SnackBar from 'atoms/SnackBar';

interface State {
  isOpen: boolean;
  message: string;
}

interface SnackbarProviderProps {
  children: React.ReactNode;
  value?: State;
}
interface SnackbarContextData extends State {
  showSnackbar: (message: string) => void;
  hideSnackBar: () => void;
}

const defaultState: State = {
  isOpen: false,
  message: '',
};

const SnackbarContext = React.createContext<SnackbarContextData | undefined>(
  undefined
);

const SnackbarProvider = ({
  children,
  value,
}: SnackbarProviderProps): JSX.Element => {
  const [{ isOpen, message }, setSnackBarState] = useState<State>(
    value || defaultState
  );

  const showSnackbar = (message: string) =>
    setSnackBarState(state => ({ ...state, isOpen: true, message }));

  const hideSnackBar = () =>
    setSnackBarState(state => ({ ...state, isOpen: false }));

  return (
    <SnackbarContext.Provider
      value={{
        isOpen,
        message,
        showSnackbar,
        hideSnackBar,
      }}>
      <SnackBar isOpen={isOpen} message={message} onClose={hideSnackBar} />
      {children}
    </SnackbarContext.Provider>
  );
};

const useSnackBar = () => {
  const context = useContext(SnackbarContext);

  if (context === undefined) {
    throw new Error(`useSnackBar must be used within a SnackbarContext`);
  }
  return context;
};

export { SnackbarProvider, useSnackBar };
