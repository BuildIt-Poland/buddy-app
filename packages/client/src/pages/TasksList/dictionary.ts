interface TaskListDictionary {
  BUDDY_TAB_TITLE: string;
  NEWBIE_TAB_TITLE: string;
  SUCCESS_MESSAGE: string;
  ERROR_MESSAGE: string;
  TEMPLATE_MESSAGE: string;
  PLUS_BUTTON_TITLE: string;
}

const TASK_LIST_DICTIONARY: TaskListDictionary = {
  BUDDY_TAB_TITLE: 'My tasks',
  NEWBIE_TAB_TITLE: 'Newbie tasks',
  SUCCESS_MESSAGE: 'Task status updated',
  ERROR_MESSAGE: 'An error ocurred updating task',
  TEMPLATE_MESSAGE: 'Task status is static for template tasks',
  PLUS_BUTTON_TITLE: 'Add new task',
};

export default TASK_LIST_DICTIONARY;
