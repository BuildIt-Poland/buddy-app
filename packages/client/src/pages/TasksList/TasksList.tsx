import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import TabPanel from 'atoms/TabPanel';
import AvatarHeader from 'components/AvatarHeader';
import { useQuery } from '@apollo/react-hooks';
import { useSnackBar } from 'contexts/SnackbarContext';
import { QueryNewbieArgs, Query, Task } from '@buddy-app/schema';
import { TASK_LIST } from 'graphql/task-list.graphql';
import { useParams, useLocation } from 'react-router-dom';
import Box from '@material-ui/core/Box';
import TaskTabsContent from 'components/TaskTabsContent';
import { ROUTES } from 'shared/routes';
import useTaskProgress from 'hooks/useTaskProgress';
import useTaskStatusUpdate from 'hooks/useTaskStatusUpdate';
import Header, { MenuTypes, MenuColors } from 'components/Header';
import AddTaskOptions from 'components/AddTaskOptions';
import DICTIONARY from './dictionary';

const TasksList: React.FC = () => {
  const { newbieId } = useParams<QueryNewbieArgs>();
  const { state } = useLocation();
  const history = useHistory();
  const { showSnackbar } = useSnackBar();
  const defaultTabIndex = (state && state.defaultTabIndex) || 0;
  const [tabIndex, setTabIndex] = useState(defaultTabIndex);

  const { loading, data } = useQuery<Query, QueryNewbieArgs>(TASK_LIST, {
    variables: { newbieId },
  });
  const { buddyProgress } = useTaskProgress(data && data.newbie);
  const [updateTaskStatus] = useTaskStatusUpdate(newbieId, {
    onCompleted: () => showSnackbar(DICTIONARY.SUCCESS_MESSAGE),
    onError: () => showSnackbar(DICTIONARY.ERROR_MESSAGE),
  });

  const handleTabChange = (event: React.ChangeEvent<{}>, newValue: number) =>
    setTabIndex(newValue);

  const onBackClick = () => {
    history.push(ROUTES.BUDDY_SELECT_NEWBIE);
  };

  const newbieTasks = data && data.newbie.newbieTasks;
  const buddyTasks = data && data.newbie.buddyTasks;
  const pathname = ROUTES.BUDDY_ADD_TASK.replace(':newbieId', newbieId);

  return (
    <>
      <Header
        type={MenuTypes.BACK}
        color={MenuColors.PAPER}
        onButtonClick={onBackClick}>
        <AvatarHeader newbieId={newbieId} taskProgress={buddyProgress} />
        <Tabs
          centered
          value={tabIndex}
          onChange={handleTabChange}
          indicatorColor='primary'
          textColor='primary'
          variant='fullWidth'>
          <Tab label={DICTIONARY.NEWBIE_TAB_TITLE} />
          <Tab label={DICTIONARY.BUDDY_TAB_TITLE} />
        </Tabs>
      </Header>
      <Box component={'main'} flex={1} data-testid='task-list-page'>
        <TabPanel value={tabIndex} index={0}>
          <TaskTabsContent
            loading={loading}
            tabIndex={tabIndex}
            onChange={updateTaskStatus}
            tasks={newbieTasks as Task[]}
          />
        </TabPanel>
        <TabPanel value={tabIndex} index={1}>
          <TaskTabsContent
            tabIndex={tabIndex}
            loading={loading}
            onChange={updateTaskStatus}
            tasks={buddyTasks as Task[]}
          />
        </TabPanel>
        <AddTaskOptions
          title={DICTIONARY.PLUS_BUTTON_TITLE}
          to={{ pathname, state: { tabIndex } }}
        />
      </Box>
    </>
  );
};
export default TasksList;
