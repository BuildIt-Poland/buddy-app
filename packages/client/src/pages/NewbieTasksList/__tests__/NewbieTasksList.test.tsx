import React from 'react';
import { create, act } from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';
import { MemoryRouter, Route } from 'react-router';
import { MockedProvider } from '@apollo/react-testing';
import {
  taskListResponse,
  newbieTasksListMock,
  mockedNewbieContext,
} from '__mocks__';
import { AuthProvider } from 'contexts/AuthContext';
import { MenuProvider } from 'contexts/MenuContext';
import { SnackbarProvider } from 'contexts/SnackbarContext';

import TasksList from '../NewbieTasksList';

jest.mock('components/AvatarHeader', () => 'AvatarHeader');
jest.mock('components/TaskTabsContent', () => 'TaskTabsContent');
jest.doMock('components/Header');

describe('Component - TasksList', () => {
  const path = '/newbie/tasks';
  describe('When there are tasks', () => {
    const variables = {
      newbieId: '1234',
    };

    test('renders correctly', async () => {
      const component = create(
        <MockedProvider
          mocks={taskListResponse(variables, newbieTasksListMock)}
          addTypename={false}
          resolvers={{}}>
          <MemoryRouter initialEntries={[path]}>
            <AuthProvider value={mockedNewbieContext()}>
              <SnackbarProvider>
                <MenuProvider>
                  <Route path={path} component={TasksList} />
                </MenuProvider>
              </SnackbarProvider>
            </AuthProvider>
          </MemoryRouter>
        </MockedProvider>
      );

      expect(component.toJSON()).toMatchSnapshot();

      await act(async () => {
        await waitForExpect(() => {
          expect(component.toJSON()).toMatchSnapshot();
        });
      });
    });
  });

  describe('when there are not tasks', () => {
    const variables = {
      newbieId: '1234',
    };

    const newbieMockData = {
      id: '1',
      name: '',
      email: '',
      newbieTasks: [],
      buddyTasks: [],
    };

    test('should render empty state', async () => {
      const component = create(
        <MockedProvider
          mocks={taskListResponse(variables, newbieMockData)}
          addTypename={false}
          resolvers={{}}>
          <MemoryRouter initialEntries={[path]}>
            <AuthProvider value={mockedNewbieContext()}>
              <SnackbarProvider>
                <MenuProvider>
                  <Route path={path} component={TasksList} />
                </MenuProvider>
              </SnackbarProvider>
            </AuthProvider>
          </MemoryRouter>
        </MockedProvider>
      );

      await act(async () => {
        await waitForExpect(() => {
          expect(component.toJSON()).toMatchSnapshot();
        });
      });
    });
  });
});
