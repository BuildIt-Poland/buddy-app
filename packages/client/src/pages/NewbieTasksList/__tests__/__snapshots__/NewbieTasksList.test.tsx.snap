// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - TasksList When there are tasks renders correctly 1`] = `
Array [
  <header
    color="paper"
    shape="rounded"
    type="menu"
  >
    <AvatarHeader
      newbieId="1234"
      taskProgress={0}
    />
  </header>,
  <main
    className="MuiBox-root MuiBox-root-9"
    data-testid="task-list-page"
  >
    <TaskTabsContent
      loading={true}
      onChange={[Function]}
    />
  </main>,
]
`;

exports[`Component - TasksList When there are tasks renders correctly 2`] = `
Array [
  <header
    color="paper"
    shape="rounded"
    type="menu"
  >
    <AvatarHeader
      newbieId="1234"
      taskProgress={1}
    />
  </header>,
  <main
    className="MuiBox-root MuiBox-root-9"
    data-testid="task-list-page"
  >
    <TaskTabsContent
      loading={false}
      onChange={[Function]}
      tasks={
        Array [
          Object {
            "description": "New task description 2",
            "id": "1",
            "status": "COMPLETED",
            "title": "New task title 2",
          },
          Object {
            "description": "New task description 2",
            "id": "1",
            "status": "COMPLETED",
            "title": "New task title 2",
          },
        ]
      }
    />
  </main>,
]
`;

exports[`Component - TasksList when there are not tasks should render empty state 1`] = `
Array [
  <header
    color="paper"
    shape="rounded"
    type="menu"
  >
    <AvatarHeader
      newbieId="1234"
      taskProgress={0}
    />
  </header>,
  <main
    className="MuiBox-root MuiBox-root-10"
    data-testid="task-list-page"
  >
    <TaskTabsContent
      loading={false}
      onChange={[Function]}
      tasks={Array []}
    />
  </main>,
]
`;
